client/services/LocalStorageService.js

```diff
@@ -0,0 +1,196 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @fileoverview A service that saves student code to the browser's
+ *    localStorage.
+ */
+
+tie.factory('LocalStorageService', ['FeedbackParagraphObjectFactory',
+  'ReinforcementBulletObjectFactory',
+  function(FeedbackParagraphObjectFactory, ReinforcementBulletObjectFactory) {
+    // In some browsers, localStorage is not available and its
+    // invocation throws an error.
+    var localStorageIsAvailable = false;
+    try {
+      localStorageIsAvailable = Boolean(localStorage);
+    } catch (e) {
+      localStorageIsAvailable = false;
+    }
+
+    /**
+     * Returns the local storage key for a given question.
+     *
+     * @param {string} questionId
+     * @param {string} language
+     * @returns {string}
+     */
+    var getLocalStorageKeyForCode = function(questionId, language) {
+      return questionId + ":" + language + ":code";
+    };
+
+    /**
+     * Returns the local storage key for the feedback.
+     *
+     * @param {string} questionId
+     * @param {string} language
+     * @returns {string}
+     */
+    var getLocalStorageKeyForFeedback = function(questionId, language) {
+      return questionId + ":" + language + ":feedback";
+    };
+
+    return {
+      /**
+       * Checks if the local storage is available.
+       *
+       * @returns {boolean}
+       */
+      isAvailable: function() {
+        return localStorageIsAvailable;
+      },
+
+      /**
+       * Stores the code into local storage - so long as local storage is
+       * available.
+       *
+       * @param {string} questionId
+       * @param {string} code
+       * @param {string} language
+       */
+      storeCode: function(questionId, code, language) {
+        if (!localStorageIsAvailable) {
+          return;
+        }
+
+        // Add version to the code, currently only 1
+        // TODO(talee): Add in check for the correct version number
```

Comment:
I don't understand this TODO; please elaborate. OK to be verbose, as long as it's clear to a new dev.